src/index.js

```diff
@@ -1,5 +1,116 @@
 'use strict'
 
-const CID = require('./cid')
+const mh = require('multihashes')
+const multibase = require('multibase')
+const multicodec = require('multicodec')
+
+const codecs = require('./codecs')
+
+// CID: <mbase><version><mcodec><mhash>
+
+class CID {
+  /*
+   * if (str)
+   *   if (1st char is on multibase table) -> CID String
+   *   else -> bs58 encoded multihash
+   * else if (Buffer)
+   *   if (0 or 1) -> CID
+   *   else -> multihash
+   * else if (Number)
+   *   -> construct CID by parts
+   *
+   * ..if only JS had traits..
+   */
+  constructor (version, codec, multihash) {
+    if (typeof version === 'string') {
+      if (multibase.isEncoded(version)) { // CID String (encoded with multibase)
+        const cid = multibase.decode(version)
+        this.version = parseInt(cid.slice(0, 1).toString('hex'), 16)
+        this.codec = multicodec.getCodec(cid.slice(1))
+        this.multihash = multicodec.rmPrefix(cid.slice(1))
+      } else { // bs58 string encoded multihash
+        this.codec = 'dag-pb'
+        this.multihash = mh.fromB58String(version)
+        this.version = 0
+      }
+    } else if (Buffer.isBuffer(version)) {
+      const firstByte = version.slice(0, 1)
+      const v = parseInt(firstByte.toString('hex'), 16)
+      if (v === 0 || v === 1) { // CID
+        const cid = version
+        this.version = v
+        this.codec = multicodec.getCodec(cid.slice(1))
+        this.multihash = multicodec.rmPrefix(cid.slice(1))
+      } else { // multihash
+        this.codec = 'dag-pb'
+        this.multihash = version
+        this.version = 0
+      }
+    } else if (typeof version === 'number') {
+      //
+      // TODO validate all of these
+      //
+      this.codec = codec
+      this.version = version
+      this.multihash = multihash
+    }
```

Comment:
Please throw some kind of error on the non working cases. This is a new format and could easily lead to implicit errors if we fail silently